make_prg/utils/seq_utils.py

```diff
@@ -0,0 +1,220 @@
+
+    @classmethod
+    def get_expanded_sequences(cls, sequences: List[str]) -> Sequences:
+        """
+        Expand sequences in the given list of sequences, following the translation table in SequenceExpander.iupac.
+        It does the following steps:
+            1. Check that we don't have disallowed bases;
+            2. Remove sequences with N;
+            3. Duplicate sequences containing RYKMSW, replacing with AGCT alternatives;
+        Note 1: The sequences are deliberately returned in the order they are received.
+        Note 2: Returned sequences are composed of ACGT only
+        """
+
+        cls.check_if_there_is_sequence_with_disallowed_bases(sequences)
+
+        expanded_seqs = []
+        expanded_set = set()
+        deduplicated_gapless_seqs = remove_duplicates(sequences)
+        for seq in deduplicated_gapless_seqs:
+            if "N" in seq:
+                continue
+
+            alternatives = [cls.iupac[base] for base in seq]
+            for tuple_product in itertools.product(*alternatives):
+                expanded_str = "".join(tuple_product)
+                if expanded_str not in expanded_set:
+                    expanded_set.add(expanded_str)
+                    expanded_seqs.append(expanded_str)
+
+        all_sequences_contained_N = len(expanded_seqs) == 0
+        if all_sequences_contained_N:
+            raise SequenceCurationError(
+                "All sequences in this slice contained N. Redo sequence curation.\n"
+                f"Sequences: {sequences}"
+            )
+
+        cls.check_all_sequences_are_composed_of_ACGT(expanded_seqs)
+        return expanded_seqs
+
+    @classmethod
+    def get_expanded_sequences_from_MSA(cls, alignment: MSA) -> Sequences:
+        gapless_seqs = list(map(ungap, get_alignment_seqs(alignment)))
+        return cls.get_expanded_sequences(gapless_seqs)
+
+
+def align(seq1: str, seq2: str,
+          match_score: float = 2,
+          mismatch_score: float = -0.9,
+          gap_open_score: float = -1.1,
+          gap_extend_score: float = -1) -> Tuple[str, str]:
+    alignment = pairwise2.align.globalms(
+        seq1, seq2, match_score, mismatch_score, gap_open_score, gap_extend_score, one_alignment_only=True
+    )
+    empty_alignment = len(alignment) == 0
+    if empty_alignment:
+        #  this only happens if ref or alt are empty
+        ref_is_empty = len(seq1) == 0
+        alt_is_empty = len(seq2) == 0
+
+        both_are_not_empty = (not ref_is_empty) and (not alt_is_empty)
+        assert not both_are_not_empty, "Alignment error: got an empty alignment from two non-empty input sequences"
+
+        if ref_is_empty:
+            return GAP * len(seq2), seq2
+        else:  # alt_is_empty
+            return seq1, GAP * len(seq1)
+    else:
+        alignment = alignment[0]
+        return alignment.seqA, alignment.seqB
+
+
+def remove_columns_full_of_gaps_from_MSA(alignment: MSA) -> MSA:
```

Comment:
Just checking, where is this function used? I'm guessing in later PR code

Reply:
Yes, in `RecursiveTreeNode.__init__()`, but I think it might be present in the current code also? It removes columns in the MSA that are full of gaps, like this second column here:

```
A-T
A-C
A-G
```

which can appear when we get sub-alignments from the full MSA.

Reply:
i couldn't find it being used here but it's fine it made sense to me